File: CreateBookFunction/src/main/java/com/mishenev/post_book/db/BookRepositoryJdbcImpl.java
package com.mishenev.post_book.db;

import com.mishenev.post_book.CreateBookRequest;

import java.io.Closeable;
import java.io.IOException;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.SQLException;

/**
 * BookRepositoryImpl.
 *
 * @author Dmitrii_Mishenev
 */
public class BookRepositoryJdbcImpl implements BookRepository, Closeable {

    private final Connection connection;

    public BookRepositoryJdbcImpl(Connection connection) {
        this.connection = connection;
    }

    public BookRepositoryJdbcImpl() {
        this.connection = new JdbcConnectionFactory().createConnection();
    }

    @Override
    public void createBook(CreateBookRequest createBookRequest) {
        final String sqlStatement = "INSERT INTO books (tittle, name, description) VALUES (?, ?, ?)";

        try(PreparedStatement insertBookStatement = connection.prepareStatement(sqlStatement)) {
            insertBookStatement.setString(1, createBookRequest.getTittle());
            insertBookStatement.setString(2, createBookRequest.getName());
            insertBookStatement.setString(3, createBookRequest.getDescription());
            insertBookStatement.execute();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void close() throws IOException {
        try {
            connection.close();
        } catch (SQLException e) {
            throw new IOException(e);
        }
    }
}

File: BooksByNameCounterFunction/src/main/java/com/mishenev/count_book/db/BookRepositoryJdbcImpl.kt
package com.mishenev.count_book.db

import java.io.Closeable
import java.io.IOException
import java.sql.Connection
import java.sql.SQLException

class BookRepositoryJdbcImpl : BookRepository, Closeable {
    private val connection: Connection

    constructor(connection: Connection) {
        this.connection = connection
    }

    constructor() {
        connection = JdbcConnectionFactory().createConnection()
    }

    override fun countBooksByName(name: String): Long {
        val sqlStatement = "SELECT COUNT(*)  AS recordCount FROM books WHERE name = ?"

         try {
            connection.prepareStatement(sqlStatement)
                .use { insertBookStatement ->
                    insertBookStatement.setString(1, name)
                    val resultSet = insertBookStatement.executeQuery()
                    resultSet.next()
                    return resultSet.getLong("recordCount")
                }
        } catch (e: SQLException) {
            throw RuntimeException(e)
        }
    }

    @Throws(IOException::class)
    override fun close() {
        try {
            connection.close()
        } catch (e: SQLException) {
            throw IOException(e)
        }
    }
}